### ISSUE TITLE
Performance: Optimize WhichMut trainer - 100-1000x speedup needed

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/netam#159]
# Performance Optimization: WhichMut Trainer

## 🚨 Critical Performance Issue

The current WhichMut trainer implementation in `netam/whichmut_trainer.py` has **severe performance bottlenecks** that make it unusable for production datasets:

- **Pipeline batches (128 samples)**: ~34 minutes per batch → **68 minutes per epoch**
- **Simple script (10 samples, batch size 2)**: ~6 minutes total
- **Root cause**: Triple nested loops computing normalization constants

## 📊 Performance Analysis

**Current Bottleneck Breakdown:**
- `compute_normalization_constants()`: **99.97%** of computation time (606M operations)
- `compute_whichmut_loss_batch()`: **0.03%** of computation time (163K operations)

**Triple Nested Loop Problem:**
```python
for seq_idx in range(N):                    # 128 sequences
    for codon_pos in range(L_codon):          # 1273 codons
        for parent_codon_idx in valid_codons:   # ~61 valid codons
            for child_codon_idx in valid_codons:  # ~61 valid codons
```
= **606,314,624 operations per batch**

## 🔧 Proposed Solution

Replace nested loops with **vectorized PyTorch operations** for 100-1000x speedup:

### 1. Vectorize `compute_normalization_constants()`
- **Current**: Triple nested loops (N × L_codon × 61 × 61)
- **Optimized**: Matrix operations with broadcasting and einsum
- **Expected speedup**: 100-1000x

### 2. Vectorize `compute_whichmut_loss_batch()`  
- **Current**: Nested loops over sequences and positions
- **Optimized**: Boolean indexing with `torch.gather()` and `torch.nonzero()`
- **Expected speedup**: 10-100x

### 3. Pre-compute Static Mappings
- Convert `AA_IDX_FROM_CODON_IDX` dictionary → tensor
- Create boolean masks for valid codons
- Cache intermediate results

## 📈 Expected Impact

- **Training time**: 68 minutes → **4 minutes per epoch** (17x speedup)
- **Full pipeline**: 2.3 hours → **8 minutes** for 2 epochs
- **Scalability**: Enables larger batch sizes and datasets
- **Resource efficiency**: Better GPU/CPU utilization

### Phase 1: Analysis & Validation (1-2 days)
1. Add performance analysis script to `netam/benchmarks/` directory
2. Profile current implementation with torch.profiler to confirm bottlenecks
3. Create unit tests for current behavior (preserve correctness)

### Phase 2: Core Optimization (3-5 days)
1. **Implement vectorized `compute_normalization_constants()`**
   - Replace triple nested loops with einsum operations
   - Add codon mapping precomputation
   - Test correctness against original implementation

2. **Implement vectorized `compute_whichmut_loss_batch()`**
   - Replace nested loops with boolean indexing
   - Use torch.gather() for efficient data access
   - Maintain gradient flow for training

### Phase 3: Integration & Testing (2-3 days)
1. Backward compatibility - keep original functions for fallback
2. Integration testing with full training pipeline
3. Performance benchmarking on various batch sizes
4. Memory usage analysis and optimization

### Phase 4: Production Deployment (1-2 days)
1. Documentation updates - function docstrings and examples
2. Configuration option to switch between implementations
3. Performance monitoring in production pipelines

## 🔄 Success Criteria

- [ ] Vectorized functions pass all existing unit tests
- [ ] 100x+ speedup for normalization constants computation
- [ ] 10x+ speedup for overall batch loss computation  
- [ ] Memory usage remains reasonable (< 2x increase)
- [ ] Gradient flow preserved for training
- [ ] Production pipeline integration successful

---

**Priority**: 🔥 **Critical** - Current implementation is unusable for production
**Effort**: ~1-2 weeks development + testing
**Impact**: Enables WhichMut training for real datasets

**Comment by matsen** [comment]
Love it!

1. I'm sure you are going to do this, but we want to have a unit test that runs things the slow-and-clear way and compares to the magic way.
2. Can you explain to me why we are looping over all parent codons in `for parent_codon_idx in valid_codons:   # ~61 valid codons`? I think I've forgotten.

**Comment by jgallowa07** [comment]
Closing this as our bug fixes have sped things up quite a bit. May address optimizations still over in #158
